node/src/components/consensus/highway_core/state.rs

```diff
@@ -705,6 +695,120 @@ impl<C: Context> State<C> {
     pub(crate) fn is_empty(&self) -> bool {
         self.units.is_empty()
     }
+
+    /// Validates whether `wunit` violates the Limited Naïveté Criterion (LNC).
+    /// Returns index of the first equivocator that was cited naively in violation of the LNC, or
+    /// `None` if the LNC is satisfied.
+    fn validate_lnc(&self, wunit: &WireUnit<C>) -> Option<ValidatorIndex> {
+        wunit
+            .panorama
+            .enumerate()
+            .filter(|(_, obs)| obs.is_faulty())
+            .map(|(i, _)| i)
+            .find(|eq_idx| !self.satisfies_lnc_for(wunit, *eq_idx))
+    }
+
+    /// Returns `true` if there is at most one fork by the validator `eq_idx` that is cited naively
+    /// by `wunit` or earlier units by the same creator.
+    fn satisfies_lnc_for(&self, wunit: &WireUnit<C>, eq_idx: ValidatorIndex) -> bool {
+        // Find all forks by eq_idx that are cited naively by wunit.
+        // * If it's more than one, return false (the LNC is violated).
+        // * If it's none, return true: If the LNC were violated we would have already noticed when
+        //   validating earlier units by wunit.creator.
+        // * Otherwise store the unique naively cited fork in naive_by_wunit.
+        let mut opt_naive_by_wunit = None;
+        {
+            let seen_by_endorsed =
+                |hash| wunit.endorsed.iter().any(|e_hash| self.sees(e_hash, hash));
+            let mut to_visit: Vec<_> = wunit.panorama.iter_correct_hashes().collect();
+            let mut added_to_to_visit: HashSet<_> = to_visit.iter().cloned().collect();
+            while let Some(hash) = to_visit.pop() {
+                if seen_by_endorsed(hash) {
+                    continue;
+                }
```

Comment:
What if Alice creates two units that do not cite each other: `a0` and `a0'`. She sends `a0` to Bob and `a0'` to Charlie. Bob endorses `a0` (and creates `b0` citing `a0`), Charlie endorses `a0'` (and creates `c0` citing `c0'`). Now, it's possible for Dan to create a unit that cites both `b0` and `c0`. When another validator receives that unit, it will check that `unit.endorsed.sees(b0)` and `unit.endorsed.sees(c0)` and we will never find that `a0` and `a0'` were an equivocation.

Am I right or I am completely wrong?

Reply:
I guess we should already have a unit test for that :thinking:

Reply:
When you say "Bob endorses `a0`", do you mean Bob gets >50% of validators to create an endorsement for `a0`, and then creates `b0` claiming that `a0` is endorsed? And Charlie manages to do the same with `a0'`? (Because there are faulty validators endorsing both of Alice's forks?)

Anyway, when Dan creates a unit that cites both `b0` and `c0`, he has to put `Faulty` into the panorama for Alice; also, everyone who just downloads `a0` and `a0'` will know that she is faulty. So I'm not sure I understand your concern.

That `validate_lnc` never considers a citation of `a0` or `a0'` to be naive is correct: Those are non-naive citations.

Reply:
> When you say "Bob endorses a0", do you mean Bob gets >50% of validators to create an endorsement for a0, and then creates b0 claiming that a0 is endorsed? And Charlie manages to do the same with a0'? (Because there are faulty validators endorsing both of Alice's forks?)

No, just that when it sees `a0` , Alice is honest according to his knowledge and he will include `a0` in its panorama and add it to `unit.claimed_endorsements`.

> That `validate_lnc` never considers a citation of a0 or a0' to be naive is correct: Those are non-naive citations.

I guess I don't understand the algorithm yet. I thought that `validate_lnc` would see `b0` and `c0` in `claimed_endorsed` so `seen_by_endorsed` will also return `true`.

Reply:
> No, just that when it sees `a0` , Alice is honest according to his knowledge and he will include a0 in its panorama and add it to `unit.claimed_endorsements`.

But when bob sees Alice as honest and doesn't cite her equivocation, he doesn't need to claim `a0` is endorsed. And if he doesn't see `a0` as endorsed, he mustn't claim it. He can just naively cite `a0`.

When `validate_lnc` is applied to Charlie's message, it sees in `claimed_endorsed` whatever Charlie claimed was endorsed. In your example, which messages would that be? If it's `[b0, c0]`, the LNC is satisfied. If it's `[a0, a0_prime]`, it's also satisfied.